feat: lua functions and closure support for '*expr'/'*func' options

famiu opened the pull request:
Adds support for Lua functions and Vim script closures for `*expr`/`*func` options. (eg: `'indentexpr'`. `'tagfunc'`, etc.)

Very WIP and currently in a broken and uncompilable state.

zeertzjq commented on src/nvim/eval/typval_defs.h:
```diff
@@ -0,0 +1,396 @@
+  ListLogEntry entries[];  ///< Actual log entries.
+};
+
+extern ListLog *list_log_first;  ///< First list log chunk, NULL if missing
+extern ListLog *list_log_last;  ///< Last list log chunk
```
These two should be in `typval.h`

lewis6991 commented on src/nvim/eval.c:
```diff
@@ -3676,37 +3677,22 @@ int get_option_tv(const char **const arg, typval_T *const rettv, const bool eval
+    Error err;
+    object_to_vim(val, rettv, &err);
+    assert(err.type == kErrorTypeNone);
+  } else if (working && hidden) {
```
This is me being picky, but I think refactors like this deserve their own PR which should be merged separately.

famiu replied:
This is actually necessary for the PR

lewis6991 replied:
I only mentioned because we specifically talked about this on matrix.

This is a big PR and will be hard to review, so even though it is unfinished, you should appreciate early feedback.

famiu committed:
feat: lua functions and closure support for `'*expr'`/`'*func'` options.

Adds support for Lua functions and Vim script closures for `*expr`/`*func` options. (eg: `'indentexpr'`. `'tagfunc'`, etc.)
